bpf/mb_connect.c

```diff
@@ -1,435 +1,180 @@
+    {"cgroup", 'c', "/sys/fs/cgroup", 0, "cgroup path"},
+    {"bpffs", 'b', "/sys/fs/bpf", 0, "BPF filesystem path"},
+    {},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+    struct env *env = state->input;
+
+    switch (key) {
+    case 'v':
+        env->verbose = true;
+        break;
+    case 'c':
+        env->cgroups_path = arg;
+        break;
+    case 'b':
+        env->bpffs = arg;
+        break;
+    case ARGP_KEY_ARG:
+        argp_usage(state);
+        break;
+    default:
+        return ARGP_ERR_UNKNOWN;
+    }
+    return 0;
+}
 
-Licensed under the Apache License, Version 2.0 (the "License");
-you may not use this file except in compliance with the License.
-You may obtain a copy of the License at
+static const struct argp argp = {
+    .options = opts,
+    .parser = parse_arg,
+    .doc = argp_program_doc,
+};
 
-    http://www.apache.org/licenses/LICENSE-2.0
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format,
+                           va_list args)
+{
+    if (level == LIBBPF_DEBUG && !env.verbose)
+        return 0;
 
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
-#include "headers/helpers.h"
-#include "headers/maps.h"
-#include "headers/mesh.h"
-#include <linux/bpf.h>
-#include <linux/in.h>
+    return vfprintf(stderr, format, args);
+}
 
-#if ENABLE_IPV4
-static __u32 outip = 1;
+static volatile bool exiting = false;
 
-static inline int udp_connect4(struct bpf_sock_addr *ctx)
-{
-#if MESH != ISTIO && MESH != KUMA
-    // only works on istio and kuma
-    return 1;
-#endif
-    if (bpf_htons(ctx->user_port) != 53) {
-        return 1;
-    }
-    if (!(is_port_listen_current_ns(ctx, ip_zero, OUT_REDIRECT_PORT) &&
-          is_port_listen_udp_current_ns(ctx, localhost, DNS_CAPTURE_PORT))) {
-        // this query is not from mesh injected pod, or DNS CAPTURE not enabled.
-        // we do nothing.
-        return 1;
-    }
+static void sig_handler(int sig) { exiting = true; }
 
-    __u64 uid = bpf_get_current_uid_gid() & 0xffffffff;
-    if (uid != SIDECAR_USER_ID) {
-        // needs rewrite
-        struct origin_info origin;
-        memset(&origin, 0, sizeof(origin));
-        set_ipv4(origin.ip, ctx->user_ip4);
-        origin.port = ctx->user_port;
-        // save original dst
-        __u64 cookie = bpf_get_socket_cookie_addr(ctx);
-        if (bpf_map_update_elem(&cookie_original_dst, &cookie, &origin,
-                                BPF_ANY)) {
-            printk("update origin cookie failed: %d", cookie);
-        }
-        ctx->user_port = bpf_htons(DNS_CAPTURE_PORT);
-        ctx->user_ip4 = localhost;
-    }
-    return 1;
+void print_env_maybe()
```

Comment:
Maybe we can move it to a separate header file somehow and make it common?

Reply:
I would love to, but so far I'm not proficient enough in `c` to dynamically iterate over structs to log the keys and values :-D